codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/DocumentationConverter.java

```diff
@@ -72,28 +71,65 @@ public static String convert(String docs) {
         htmlDocs = Jsoup.clean(htmlDocs, GODOC_ALLOWLIST);
 
         // Now we parse the html and visit the resultant nodes to render the godoc.
-        FormattingVisitor formatter = new FormattingVisitor();
+        FormattingVisitor formatter = new FormattingVisitor(docWrapLength);
         Node body = Jsoup.parse(htmlDocs).body();
-        NodeTraversor.traverse(formatter, body);
+        traverse(formatter, body);
         return formatter.toString();
     }
 
+    private static void traverse(NodeVisitor visitor, Node root) {
+        Node node = root;
+        Node parent; // remember parent to find nodes that get replaced in .head
+        int depth = 0;
+
+        while (node != null) {
+            parent = node.parentNode();
+            visitor.head(node, depth); // visit current node
+            if (parent != null && !node.hasParent()) { // must have been replaced; find replacement
+                node = parent.childNode(node.siblingIndex()); // replace ditches parent but keeps sibling index
+            }
+
+            if (node.childNodeSize() > 0) { // descend
+                node = node.childNode(0);
+                depth++;
+            } else {
+                while (true) {
```

Comment:
Endless loops with break conditions are error-prone, especially here where we're operating on a foreign structure. Ideally you'd write this in a way where you're not liable to get stuck during iteration. This is another reason tree traversal lends itself to recursion.